File: cses/07_mathematics/1730_nim-game-i/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần cho nhanh
    data = sys.stdin.buffer.read().split()
    idx = 0
    t = int(data[idx]); idx += 1
    out = []
    for _ in range(t):
        n = int(data[idx]); idx += 1
        # Tính Nim sum: XOR dồn kích thước tất cả các đống
        x = 0
        for j in range(idx, idx + n):
            x ^= int(data[j])
        idx += n
        # Bouton's theorem: XOR != 0 thì người đi trước thắng, ngược lại người đi sau thắng
        out.append("first" if x != 0 else "second")
    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()

File: cses/07_mathematics/1730_nim-game-i/solution.cpp
#include <iostream>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int tests;
    std::cin >> tests;
    while (tests-- > 0) {
        int n;
        std::cin >> n;
        // Nim sum: XOR dồn kích thước tất cả các đống
        long long nim_sum = 0;
        for (int i = 0; i < n; ++i) {
            long long heap;
            std::cin >> heap;
            nim_sum ^= heap;
        }
        // Bouton's theorem: XOR != 0 thì người đi trước thắng, ngược lại người đi sau thắng
        std::cout << (nim_sum != 0 ? "first" : "second") << '\n';
    }
    return 0;
}
